charm/src/charm.py

```diff
@@ -0,0 +1,186 @@
+            self.mongodb.on.endpoints_changed,
+            self._on_mongodb_client_relation_changed,
+        )
+        self.framework.observe(
+            self.on.mongodb_client_relation_broken,
+            self._on_mongodb_client_relation_removed,
+        )
+        self.framework.observe(
+            self.on.testflinger_pebble_ready, self._on_testflinger_pebble_ready
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+
+    @property
+    def version(self) -> str:
+        """Report the current version of the app"""
+        # TODO: get the version somehow and return it
+        return "Version ?"
+
+    def _on_testflinger_pebble_ready(
+        self, event: ops.PebbleReadyEvent
+    ) -> None:
+        """Handle pebble-ready event."""
+        container = event.workload
+        container.add_layer("testflinger", self._pebble_layer, combine=True)
+        container.replan()
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_mongodb_client_relation_changed(
+        self, event: DatabaseCreatedEvent
+    ) -> None:
+        """Event is fired when mongodb database is created."""
+        if "mongodb" not in self._stored.reldata:
+            self._stored.reldata["mongodb"] = {}
+
+        initial = dict(self._stored.reldata["mongodb"])
+        self._stored.reldata["mongodb"].update(
+            self.mongodb.fetch_relation_data()[event.relation.id]
+        )
+        if initial != self._stored.reldata["mongodb"]:
+            self._update_layer_and_restart(None)
+
+    def _update_layer_and_restart(self, event) -> None:
+        """Define and start layer for testflinger using Pebble"""
+
+        if not self.container.can_connect():
+            self.unit.status = ops.WaitingStatus(
+                "Waiting for Pebble in workload container"
+            )
+            return
+
+        self.unit.status = ops.MaintenanceStatus("Assembling pod spec")
+        new_layer = self._pebble_layer.to_dict()
+
+        # Get the current pebble layer config
+        services = self.container.get_plan().to_dict().get("services", {})
+        if services != new_layer["services"]:
+            # Changes were made, add the new layer
+            self.container.add_layer(
+                "testflinger", self._pebble_layer, combine=True
+            )
+            logger.info("Added updated layer 'testflinger' to Pebble plan")
+
+            self.container.restart(self.pebble_service_name)
+            logger.info("Restarted '%s' service", self.pebble_service_name)
+
+        # add workload version in juju status
+        self.unit.set_workload_version(self.version)
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_mongodb_client_relation_removed(
+        self, event: ops.framework.EventBase
+    ) -> None:
+        """Event is fired when relation with mongodb is broken."""
+        self.unit.status = ops.WaitingStatus("Waiting for database relation")
+        raise SystemExit(0)
```

Comment:
`sys.exit(0)` is a preferable way to raise this exception. See https://docs.python.org/3/library/exceptions.html#SystemExit

Reply:
done